import data analysis libraries

personal data tool lib

In [1]:
import importlib.util
import sys
from pathlib import Path

# CHANGE THIS PATH to where dpf.py actually lives
dpf_path = Path("/home/theodorescottwillis/Documents/GitHub/SECOM-Process-Sensor-Analysis/dpf.py")

spec = importlib.util.spec_from_file_location("dpf", dpf_path)
dpf = importlib.util.module_from_spec(spec)
sys.modules["dpf"] = dpf
spec.loader.exec_module(dpf)

# Now test
dpf.Check


<function dpf.Check(df)>

other data analysis libraries

In [2]:
import pandas as pd
import numpy as np
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
import matplotlib.pyplot as plt
import seaborn as sns


Load dataset

In [3]:
df = pd.read_csv('secom_data.csv')
print("Original dataset shape:", df.shape)

Original dataset shape: (1567, 592)


Separate features and target (Pass/Fail)

In [4]:
target_col = 'Pass/Fail'
y = df[target_col]
X = df.drop(columns=[target_col])

print("Features shape after dropping target column from X:", X.shape)

Features shape after dropping target column from X: (1567, 591)


Drop non-predictive columns

In [5]:
# Drop timestamp column
if 'Timestamp' in X.columns:
    X = X.drop(columns=['Timestamp'])

print("Features shape after dropping timestamp column:", X.shape)

Features shape after dropping timestamp column: (1567, 590)


Remove constant sensors

In [6]:
# Threshold: unique values <= 1
constant_cols = [c for c in X.columns if X[c].nunique() <= 1]
print(f"Dropping {len(constant_cols)} constant features: {constant_cols[:5]} ...")
X = X.drop(columns=constant_cols)
print("Shape after removing constant features:", X.shape)

Dropping 116 constant features: ['Feature_6', 'Feature_14', 'Feature_43', 'Feature_50', 'Feature_53'] ...
Shape after removing constant features: (1567, 474)


Handle missing values

In [7]:
# Use median imputation (robust to outliers, but maybe we should remove outliers?)
imputer = SimpleImputer(strategy='median')
X_imputed = pd.DataFrame(imputer.fit_transform(X), columns=X.columns)
print("Missing values after imputation:", X_imputed.isnull().sum().sum())

Missing values after imputation: 0


Standardize features

In [8]:
# standardization
scaler = StandardScaler()
X_scaled = pd.DataFrame(scaler.fit_transform(X_imputed), columns=X_imputed.columns)
print("Feature scaling complete. Sample values:")
display(X_scaled.iloc[:5, :5])

Feature scaling complete. Sample values:


Feature_1  Feature_2  Feature_3  Feature_4  Feature_5
0   0.224463   0.849523  -0.436430   0.035804  -0.050121
1   1.107287  -0.383106   1.016977   0.155282  -0.059585
2  -1.114000   0.798901  -0.481447   0.688278  -0.047447
3  -0.350156  -0.199072  -0.051705  -1.104376  -0.050831
4   0.242296   0.087328   1.117227  -0.156616  -0.047033

Sanity check

In [9]:
dpf.Check(pd.concat([X_scaled, y], axis=1))

Initating Data Checking Process...
Shape of the DataFrame:
Shape: 1567 rows, 475 columns

               Dtype  Missing  Missing %  Unique
Feature_1    float64        0        0.0    1520
Feature_2    float64        0        0.0    1505
Feature_3    float64        0        0.0     507
Feature_4    float64        0        0.0     518
Feature_5    float64        0        0.0     503
...              ...      ...        ...     ...
Feature_587  float64        0        0.0     322
Feature_588  float64        0        0.0     260
Feature_589  float64        0        0.0     120
Feature_590  float64        0        0.0     612
Pass/Fail      int64        0        0.0       2

[475 rows x 4 columns]

First 5 rows:
   Feature_1  Feature_2  Feature_3  Feature_4  Feature_5  Feature_7  \
0   0.224463   0.849523  -0.436430   0.035804  -0.050121  -0.564354   
1   1.107287  -0.383106   1.016977   0.155282  -0.059585   0.197639   
2  -1.114000   0.798901  -0.481447   0.688278  -0.047447  -0.906768   

Dtype  Missing  Missing %  Unique
Feature_1    float64        0        0.0    1520
Feature_2    float64        0        0.0    1505
Feature_3    float64        0        0.0     507
Feature_4    float64        0        0.0     518
Feature_5    float64        0        0.0     503
...              ...      ...        ...     ...
Feature_587  float64        0        0.0     322
Feature_588  float64        0        0.0     260
Feature_589  float64        0        0.0     120
Feature_590  float64        0        0.0     612
Pass/Fail      int64        0        0.0       2

[475 rows x 4 columns]

Save the cleaned DataFrame to a CSV

In [10]:
# Save cleaned dataset (features + target) to CSV
cleaned_df = pd.concat([X_scaled, y], axis=1)
cleaned_df.to_csv('secom_cleaned.csv', index=False)
print(f"Saved cleaned data to 'secom_cleaned.csv' (shape: {cleaned_df.shape})")

Saved cleaned data to 'secom_cleaned.csv' (shape: (1567, 475))


## SECOM – Data Cleaning Phase Summary

In this phase, we performed a thorough cleaning and preliminary exploration of the SECOM dataset to prepare it for downstream analysis and modeling.

### Key Steps:

1. **Data Overview**
   - Loaded the dataset with 1,567 samples and 590 sensor features, plus the `Pass/Fail` label.
   - Initial inspection with `dpf.Check()` highlighted missing values, data types, and the number of unique values per column.

2. **Handling Missing Values**
   - Identified columns with missing values (mostly numerical sensor readings).  
   - Applied imputation strategies where appropriate (mean/median/mode) and removed rows when necessary.

3. **Final Cleaned Dataset**
   - Cleaned DataFrame contains fully imputed sensor readings with no constant columns.
   - Maintains all 1,567 observations with reduced feature redundancy.


This structured cleaning ensures the dataset is consistent, informative, and ready for analytical modeling.
